### ISSUE TITLE
Off by one error in fblas_l1.pyf.src

### ISSUE THREAD

**Issue by ewmoore** [scipy/scipy#3983]
```
import numpy as np
from scipy import linalg

a = np.arange(7, dtype='d')
b = np.zeros(4)

linalg.blas.dswap(a, b, incx=2)

np.testing.assert_allclose(b, [0,2,4,6])
```

Returns

```
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

(mismatch 100.0%)
 x: array([ 0.,  2.,  4.,  0.])
 y: array([0, 2, 4, 6])
```

AFAICT nearly all (or perhaps all) of the calculations of `n` in [fblas_l1.pyf.src](https://github.com/scipy/scipy/blob/master/scipy/linalg/fblas_l1.pyf.src) aren't quite right.  Seems like they should be 

```
n = (len(x)-1-offx)/abs(incx)+1
```

**Comment by melissawm** [comment]
Sorry about the copy-paste snafu above. Is this worth fixing? The issue is still present - I can fix and add simple tests, but I'm just wondering because so much time has passed.

**Comment by HaoZeke** [comment]
Yes it is definitely worth fixing, even if eventually the goal should be that F2PY can generate correct `pyf` files instead of needing the overhead of `src` and more.

**Comment by ilayn** [comment]
These are still obeying Fortran standards (1-indexed) and all wrappers are following the original functions. That's why there is a difference between the thin wrappers and SciPy functions where we take care of these.

**Comment by ilayn** [comment]
Ah OK, these are BLAS funcs 🤦  and still have `dimension(*)` in the wrappers. I've fixed the LAPACK ones some years ago but apparently didn't finish the BLAS part. Here, as @WarrenWeckesser found out, `n` should have been an input constrained by the length of the inputs but I'm not sure if we can fix it without breaking things.

**Comment by ev-br** [comment]
was this fixed in your no-f2py refactor @ilayn ?

**Comment by ilayn** [comment]
The `main` says no. But I can fix it with the LAPACK wrappers. I am on the last straight with the `flapack_other.pyf.src`.
